### PR TITLE
fix: Deno.symlink and Deno.symlinkSync require read and write permissions

### LINKED ISSUE #12152: [Security] Deno Sandbox Escape
The Deno file sandbox does not handle symbolic links correctly.
When running Deno with specific write access the `Deno.symlink` method can be used to gain access to any directory.

Proof of concept:
https://hackers.report/report/614876917a7b150012836bb8

### PR DESCRIPTION
**Pull request by bartlomieju**
Fixes #12152

### CODE REVIEW

**Comment by caspervonb** [comment]
Been discussing a bit offline with Bartek, re-iterating some of the main points we touched on here:
- This is a symptom of https://github.com/denoland/deno/issues/2318
- As this is a symptom, it doesn't actually fix the real problem.
- We should canonicalize when checking permissions https://github.com/denoland/deno/issues/9607 (this will take a hit on performance).
- Requiring permissions what the symlink points at is awkward because that doesn't actually have to **exist** and/or it can also be a symlink. If we require read and write on the old path and canonicalise, this gets strange as that would follow symlinks during the time of symlink creation which seems very undesirable.

**Comment by ry** [comment]
This PR fixes #2318.

If there is an existing symlink (before Deno was started) we are not going to try to guard access to that. We do need to ensure symlinks cannot be created outside of the sandbox.

**Comment by caspervonb** [comment]
> This PR fixes #2318.
> 
> If there is an existing symlink (before Deno was started) we are not going to try to guard access to that. We do need to ensure symlinks cannot be created outside of the sandbox.

It really doesn't.
All I need to get root access with this, is create a relative symlink pointing back a few levels, then move that to cwd.

**Comment by ry** [comment]
> Assuming I have access to cwd then all I need to get root access with this is to create a relative symlink pointing back a few relative levels, then move that to cwd.

You won't be able to create that symlink after this patch (if I understand you correctly)

**Comment by caspervonb** [comment]
> > Assuming I have access to cwd then all I need to get root access with this is to create a relative symlink pointing back a few relative levels, then move that to cwd.
> 
> You won't be able to create that symlink after this patch (if I understand you correctly)

One of many possibilities:

- Run deno with --allow-read=./ --allow-write=./
- Create a new directory
- Create a symlink in the new directory to ..
- Move that symlink to ./
- Escape

**Comment by ry** [comment]
> Create a symlink in the new directory to ..

That won't be allowed after this patch.

**Review by ry — APPROVED** [review]
LGTM

**Comment by caspervonb** [comment]
> > Create a symlink in the new directory to ..
> 
> That won't be allowed after this patch.

Yes it will, because if you have read-write to `./` then creating a symlink when in `./new_dir` that points to `..` is still allowed.

File: runtime/ops/fs.rs [hunk 1]
```diff
@@ -1252,6 +1252,9 @@ fn op_symlink_sync(
   let oldpath = PathBuf::from(&args.oldpath);
   let newpath = PathBuf::from(&args.newpath);
 
+  state.borrow_mut::<Permissions>().read.check(&oldpath)?;
```
Comment (piscisaureus) [inline runtime/ops/fs.rs:1255]: * This is not correct. Symlinks are resolved relative to their own location; this assumes that they're resolved relative to CWD.
* A sandbox escape is still possible by creating a symlink to "../secret" in "./allowed1/allowed2/symlink_here", and then moving the symlink from "allowed1/allowed2" to "allowed1".

**Review by piscisaureus — CHANGES_REQUESTED** [review]

File: runtime/ops/fs.rs (hunk 1 above)
Comment (bartlomieju) [reply runtime/ops/fs.rs:1255]: > This is not correct. Symlinks are resolved relative to their own location; this assumes that they're resolved relative to CWD.

Can you elaborate on this one?

> A sandbox escape is still possible by creating a symlink to "../secret" in "./allowed1/allowed2/symlink_here", and then moving the symlink from "allowed1/allowed2" to "allowed1".

Yes, @ry and I both agree on that and count it as separate issue tracked in #2318 and #9607.

The intent of this PR was to fix the immediate bug of not checking one permission when creating a symlink.

Comment (piscisaureus) [reply runtime/ops/fs.rs:1255]: > > This is not correct. Symlinks are resolved relative to their own location; this assumes that they're resolved relative to CWD.
> 
> Can you elaborate on this one?

If you create a symlink named `some/dir/over/there/mylink` with content `../target`, the symlink will point at `some/link/over/target`. So the permissions should be checked for these two paths:
  * `some/dir/over/there/mylink` (`newpath`)
  * `some/link/over/target` (`join(dirname(newpath), oldpath)`)

It currently checks permissions for:
  * `some/dir/over/there/mylink` (`newpath` - correct)
  * `../target` (`oldpath` - incorrect)

**Comment by bartlomieju** [comment]
Discussed offline with @piscisaureus, #12170 is a better solution.

**Pull request closed without merging**